pkg/service/service.go

```diff
@@ -0,0 +1,216 @@
+package service
+
+import (
+	"context"
+	"time"
+
+	"github.com/boltdb/bolt"
+	"github.com/pkg/errors"
+	"github.com/segmentio/ksuid"
+	"github.com/vmihailenco/msgpack/v5"
+
+	"github.com/dlmiddlecote/ooohh"
+)
+
+var _ ooohh.Service = &service{}
+
+type service struct {
+	db  *bolt.DB
+	now func() time.Time
+}
+
+func NewService(db *bolt.DB, now func() time.Time) (*service, error) {
+
+	// Initialize top-level buckets.
+	txn, err := db.Begin(true)
+	if err != nil {
+		return nil, errors.Wrap(err, "beginning transaction")
+	}
+	defer txn.Rollback()
```

Comment:
Does the `defer`ed function call not get called if the block returns first?

Reply:
The defer will not be called if the function returns before the `defer` keyword. It will always be called once it’s been declared. I assume this is what you’re asking?

The other thing you could be asking is, why is it ok to always call this? Essentially, the last call is ‘commit’ if all is good, so we’ll be rolling back nothing.